File: 169-majority-element/majority-element.cpp
class Solution {
public:
    // moore voting algorithm
    int majorityElement(vector<int>& nums) {

        int n = nums.size();
        int elem, count = 0;
        for (int i = 0; i < n; i++) {
            if (count == 0) {
                count = 1;
                elem = nums[i];
            } else if (nums[i] == elem) {
                count++;
            } else
                count--;
        }

        int cnt = 0;
        for (int i = 0; i < n; i++) {
            if (elem == nums[i])
                cnt++;
        }
        if (cnt > n/2)
            return elem;

        return -1;
    }
};

File: 169-majority-element/majority-element.java
class Solution {
    public int majorityElement(int[] nums) {
        int candicate = 0;
        int count = 0 ;

         for(int it : nums){
            if(count==0){
                candicate = it;
            }
            if(candicate == it) {
                count++;
            }
            else{
                 count--;
            }
         }
         return candicate ;
    }
}